File: src/backend/SunnySeat.Core/Utils/ShadowGeometry.cs
using NetTopologySuite.Geometries;
using NetTopologySuite.Algorithm;
using SunnySeat.Core.Entities;

namespace SunnySeat.Core.Utils;

/// <summary>
/// Utilities for 2.5D shadow geometry calculations
/// </summary>
public static class ShadowGeometry
{
    /// <summary>
    /// Maximum reliable shadow distance in meters (beyond this, shadows become unreliable)
    /// </summary>
    public const double MaxShadowDistance = 200.0;

    /// <summary>
    /// Minimum building height for meaningful shadow calculation (meters)
    /// </summary>
    public const double MinMeaningfulHeight = 3.0;

    /// <summary>
    /// Minimum sun elevation angle for reliable shadow calculations (degrees)
    /// </summary>
    public const double MinReliableElevation = 5.0;

    /// <summary>
    /// Project a building footprint to create shadow polygon using solar position
    /// </summary>
    /// <param name="buildingFootprint">Building footprint polygon</param>
    /// <param name="buildingHeight">Height of building in meters</param>
    /// <param name="solarPosition">Current solar position</param>
    /// <param name="geometryFactory">Geometry factory for creating polygons</param>
    /// <returns>Shadow polygon or null if no shadow cast</returns>
    public static Polygon? ProjectBuildingShadow(
        Polygon buildingFootprint, 
        double buildingHeight,
        SolarPosition solarPosition,
        GeometryFactory geometryFactory)
    {
        // Don't calculate shadow if sun is below horizon or too low
        if (solarPosition.Elevation <= 0 || solarPosition.Elevation < MinReliableElevation)
            return null;

        // Calculate shadow length based on building height and sun elevation
        var shadowLength = CalculateShadowLength(buildingHeight, solarPosition.Elevation);
        
        // Limit shadow length to reasonable maximum
        shadowLength = Math.Min(shadowLength, MaxShadowDistance);

        // Calculate shadow direction (opposite of sun azimuth)
        var shadowDirection = (solarPosition.Azimuth + 180) % 360;

        // Project building footprint in shadow direction
        return ProjectPolygonInDirection(buildingFootprint, shadowLength, shadowDirection, geometryFactory);
    }

    /// <summary>
    /// Calculate shadow length given building height and sun elevation angle
    /// </summary>
    /// <param name="buildingHeight">Height of building in meters</param>
    /// <param name="sunElevationDegrees">Sun elevation angle in degrees</param>
    /// <returns>Shadow length in meters</returns>
    public static double CalculateShadowLength(double buildingHeight, double sunElevationDegrees)
    {
        if (sunElevationDegrees <= 0)
            return 0;

        var sunElevationRadians = sunElevationDegrees * Math.PI / 180.0;
        return buildingHeight / Math.Tan(sunElevationRadians);
    }

    /// <summary>
    /// Project a polygon in a specific direction by a given distance
    /// </summary>
    /// <param name="polygon">Original polygon to project</param>
    /// <param name="distance">Distance to project in meters</param>
    /// <param name="directionDegrees">Direction in degrees (0 = North, clockwise)</param>
    /// <param name="geometryFactory">Geometry factory</param>
    /// <returns>Shadow polygon combining original and projected footprint</returns>
    public static Polygon ProjectPolygonInDirection(
        Polygon polygon, 
        double distance, 
        double directionDegrees,
        GeometryFactory geometryFactory)
    {
        // Convert direction to radians and adjust for coordinate system
        // (Geographic: 0° = North, Cartesian: 0° = East, so subtract 90° and negate)
        var directionRadians = -(directionDegrees - 90) * Math.PI / 180.0;
        
        var dx = distance * Math.Cos(directionRadians);
        var dy = distance * Math.Sin(directionRadians);

        // Convert meters to approximate degrees (rough approximation for shadow calculations)
        // At Gothenburg latitude (~58°), 1° longitude ? 55.8 km, 1° latitude ? 111.3 km
        var dxDegrees = dx / 55800.0; // meters to degrees longitude
        var dyDegrees = dy / 111300.0; // meters to degrees latitude

        // Get original coordinates
        var originalCoords = polygon.ExteriorRing.Coordinates;
        
        // Project coordinates
        var projectedCoords = originalCoords.Select(coord => 
            new Coordinate(coord.X + dxDegrees, coord.Y + dyDegrees)).ToArray();

        // Create shadow polygon by combining original and projected coordinates
        var shadowCoords = new List<Coordinate>();
        
        // Add original polygon coordinates
        shadowCoords.AddRange(originalCoords.Take(originalCoords.Length - 1)); // Exclude duplicate last coordinate
        
        // Add projected coordinates in reverse order to create proper polygon
        shadowCoords.AddRange(projectedCoords.Reverse().Take(projectedCoords.Length - 1));
        
        // Close the polygon
        shadowCoords.Add(shadowCoords[0]);

        // Create convex hull to ensure valid geometry
        var points = shadowCoords.Select(coord => geometryFactory.CreatePoint(coord)).ToArray();
        var convexHull = geometryFactory.CreateMultiPoint(points).ConvexHull();

        return convexHull as Polygon ?? geometryFactory.CreatePolygon();
    }

    /// <summary>
    /// Calculate intersection of shadow with patio and return shadow coverage percentage
    /// </summary>
    /// <param name="patioGeometry">Patio polygon geometry</param>
    /// <param name="shadowGeometry">Shadow polygon geometry</param>
    /// <returns>Percentage of patio covered by shadow (0-100)</returns>
    public static double CalculateShadowCoveragePercent(Polygon patioGeometry, Polygon shadowGeometry)
    {
        try
        {
            var intersection = patioGeometry.Intersection(shadowGeometry);
            if (intersection.IsEmpty)
                return 0.0;

            var intersectionArea = intersection.Area;
            var patioArea = patioGeometry.Area;

            return patioArea > 0 ? (intersectionArea / patioArea) * 100.0 : 0.0;
        }
        catch (Exception)
        {
            // Handle geometry calculation errors gracefully
            return 0.0;
        }
    }

    /// <summary>
    /// Calculate shadowed and sunlit geometries within patio
    /// </summary>
    /// <param name="patioGeometry">Patio polygon</param>
    /// <param name="shadowGeometries">Collection of shadow polygons</param>
    /// <param name="geometryFactory">Geometry factory</param>
    /// <returns>Tuple of (shadowed geometry, sunlit geometry)</returns>
    public static (Polygon? Shadowed, Polygon? Sunlit) CalculateShadowedAndSunlitAreas(
        Polygon patioGeometry,
        IEnumerable<Polygon> shadowGeometries,
        GeometryFactory geometryFactory)
    {
        try
        {
            // Union all shadow geometries
            Geometry? combinedShadows = null;
            foreach (var shadow in shadowGeometries)
            {
                combinedShadows = combinedShadows?.Union(shadow) ?? shadow;
            }

            if (combinedShadows == null)
            {
                // No shadows - entire patio is sunlit
                return (null, patioGeometry);
            }

            // Calculate shadowed area within patio
            var shadowedArea = patioGeometry.Intersection(combinedShadows);
            var shadowedPolygon = shadowedArea.IsEmpty ? null : 
                (shadowedArea as Polygon ?? GetLargestPolygon(shadowedArea, geometryFactory));

            // Calculate sunlit area (patio minus shadows)
            var sunlitArea = patioGeometry.Difference(combinedShadows);
            var sunlitPolygon = sunlitArea.IsEmpty ? null :
                (sunlitArea as Polygon ?? GetLargestPolygon(sunlitArea, geometryFactory));

            return (shadowedPolygon, sunlitPolygon);
        }
        catch (Exception)
        {
            // Handle geometry calculation errors - assume no shadows
            return (null, patioGeometry);
        }
    }

    /// <summary>
    /// Get the largest polygon from a potentially multi-polygon geometry
    /// </summary>
    private static Polygon? GetLargestPolygon(Geometry geometry, GeometryFactory geometryFactory)
    {
        if (geometry is Polygon polygon)
            return polygon;

        if (geometry is MultiPolygon multiPolygon)
        {
            return multiPolygon.Geometries
                .OfType<Polygon>()
                .OrderByDescending(p => p.Area)
                .FirstOrDefault();
        }

        return null;
    }

    /// <summary>
    /// Calculate confidence score for shadow calculation based on various factors
    /// </summary>
    /// <param name="building">Building casting shadow</param>
    /// <param name="solarPosition">Solar position</param>
    /// <param name="shadowLength">Calculated shadow length</param>
    /// <returns>Confidence score (0.0 to 1.0)</returns>
    public static double CalculateShadowConfidence(
        Building building, 
        SolarPosition solarPosition, 
        double shadowLength)
    {
        var confidence = 1.0;

        // Reduce confidence for low sun angles
        if (solarPosition.Elevation < 10.0)
            confidence *= 0.7;
        else if (solarPosition.Elevation < 20.0)
            confidence *= 0.9;

        // Reduce confidence for very long shadows (may be inaccurate)
        if (shadowLength > 100.0)
            confidence *= 0.8;
        else if (shadowLength > 50.0)
            confidence *= 0.9;

        // Adjust confidence based on building height data quality
        confidence *= building.HeightSource switch
        {
            HeightSource.Surveyed => 1.0,
            HeightSource.Osm => 0.85,
            HeightSource.Heuristic => 0.7,
            _ => 0.6
        };

        // Ensure confidence is within valid range
        return Math.Max(0.0, Math.Min(1.0, confidence));
    }
}

File: nextjs-app/lib/solar/shadow-geometry.ts
import * as turf from '@turf/turf';
import type { SolarPosition, Building, HeightSource } from './types';

const MAX_SHADOW_DISTANCE = 200.0;
const MIN_MEANINGFUL_HEIGHT = 3.0;
const MIN_RELIABLE_ELEVATION = 5.0;

// At Gothenburg latitude (~58°N): 1° lon ≈ 55.8 km, 1° lat ≈ 111.3 km
const METERS_PER_DEG_LON = 55800.0;
const METERS_PER_DEG_LAT = 111300.0;

export { MAX_SHADOW_DISTANCE, MIN_MEANINGFUL_HEIGHT, MIN_RELIABLE_ELEVATION };

export function projectBuildingShadow(
  buildingFootprint: GeoJSON.Polygon,
  buildingHeight: number,
  solarPosition: SolarPosition
): GeoJSON.Polygon | null {
  if (solarPosition.elevation <= 0 || solarPosition.elevation < MIN_RELIABLE_ELEVATION) {
    return null;
  }

  let shadowLength = calculateShadowLength(buildingHeight, solarPosition.elevation);
  shadowLength = Math.min(shadowLength, MAX_SHADOW_DISTANCE);

  const shadowDirection = (solarPosition.azimuth + 180) % 360;
  return projectPolygonInDirection(buildingFootprint, shadowLength, shadowDirection);
}

export function calculateShadowLength(
  buildingHeight: number,
  sunElevationDegrees: number
): number {
  if (sunElevationDegrees <= 0) return 0;
  const sunElevationRadians = sunElevationDegrees * (Math.PI / 180.0);
  return buildingHeight / Math.tan(sunElevationRadians);
}

export function projectPolygonInDirection(
  polygon: GeoJSON.Polygon,
  distance: number,
  directionDegrees: number
): GeoJSON.Polygon {
  const directionRadians = -(directionDegrees - 90) * (Math.PI / 180.0);
  const dx = distance * Math.cos(directionRadians);
  const dy = distance * Math.sin(directionRadians);

  const dxDeg = dx / METERS_PER_DEG_LON;
  const dyDeg = dy / METERS_PER_DEG_LAT;

  const ring = polygon.coordinates[0];
  const origCoords = ring.slice(0, -1);
  const projectedCoords = origCoords.map(([x, y]) => [x + dxDeg, y + dyDeg]);

  const allCoords = [
    ...origCoords,
    ...projectedCoords.slice().reverse(),
    origCoords[0],
  ];

  try {
    const multiPoint = turf.multiPoint(allCoords);
    const hull = turf.convex(multiPoint);
    if (hull) return hull.geometry as GeoJSON.Polygon;
  } catch {
    // Fall through to raw polygon
  }

  return { type: 'Polygon', coordinates: [allCoords.concat([allCoords[0]])] };
}

export function calculateShadowCoveragePercent(
  venueGeometry: GeoJSON.Polygon,
  shadowGeometry: GeoJSON.Polygon
): number {
  try {
    const venuePoly = turf.polygon(venueGeometry.coordinates);
    const shadowPoly = turf.polygon(shadowGeometry.coordinates);
    const intersection = turf.intersect(
      turf.featureCollection([venuePoly, shadowPoly])
    );

    if (!intersection) return 0.0;

    const intersectionArea = turf.area(intersection);
    const venueArea = turf.area(venuePoly);

    return venueArea > 0 ? (intersectionArea / venueArea) * 100.0 : 0.0;
  } catch {
    return 0.0;
  }
}

export function calculateShadowedAndSunlitAreas(
  venueGeometry: GeoJSON.Polygon,
  shadowGeometries: GeoJSON.Polygon[]
): { shadowed: GeoJSON.Polygon | null; sunlit: GeoJSON.Polygon | null } {
  try {
    if (shadowGeometries.length === 0) {
      return { shadowed: null, sunlit: venueGeometry };
    }

    let combinedShadows: GeoJSON.Feature<GeoJSON.Polygon | GeoJSON.MultiPolygon> | null = null;
    for (const shadow of shadowGeometries) {
      const shadowFeature = turf.polygon(shadow.coordinates);
      if (!combinedShadows) {
        combinedShadows = shadowFeature;
      } else {
        const result = turf.union(
          turf.featureCollection([combinedShadows, shadowFeature])
        );
        if (result) combinedShadows = result as GeoJSON.Feature<GeoJSON.Polygon | GeoJSON.MultiPolygon>;
      }
    }

    if (!combinedShadows) return { shadowed: null, sunlit: venueGeometry };

    const venuePoly = turf.polygon(venueGeometry.coordinates);

    const shadowedIntersection = turf.intersect(
      turf.featureCollection([venuePoly, combinedShadows])
    );
    const sunlitDifference = turf.difference(
      turf.featureCollection([venuePoly, combinedShadows])
    );

    const shadowed = shadowedIntersection
      ? extractPolygon(shadowedIntersection.geometry)
      : null;
    const sunlit = sunlitDifference
      ? extractPolygon(sunlitDifference.geometry)
      : null;

    return { shadowed, sunlit };
  } catch {
    return { shadowed: null, sunlit: venueGeometry };
  }
}

export function calculateShadowConfidence(
  building: Building,
  solarPosition: SolarPosition,
  shadowLength: number
): number {
  let confidence = clamp01(building.qualityScore ?? 0.7);

  if (solarPosition.elevation < 10.0) confidence *= 0.7;
  else if (solarPosition.elevation < 20.0) confidence *= 0.9;

  if (shadowLength > 100.0) confidence *= 0.8;
  else if (shadowLength > 50.0) confidence *= 0.9;

  const heightMultiplier: Record<HeightSource, number> = {
    Surveyed: 1.0,
    Osm: 0.85,
    Heuristic: 0.7,
    ManualOverride: 0.6,
  };
  confidence *= heightMultiplier[building.heightSource] ?? 0.6;
  confidence *= sourcePriorityMultiplier(building.sourcePriority);
  confidence *= shadowCasterTierMultiplier(building.shadowCasterTier);
  confidence *= filterDecisionMultiplier(building.filterDecision);
  confidence *= casterClassMultiplier(building.casterClass);

  return clamp01(confidence);
}

function sourcePriorityMultiplier(priority: number | undefined): number {
  if (priority === undefined || !Number.isFinite(priority)) return 0.7;
  if (priority <= 10) return 1;
  if (priority <= 30) return 0.95;
  if (priority <= 40) return 0.9;
  if (priority <= 90) return 0.75;
  return 0.65;
}

function shadowCasterTierMultiplier(tier: Building['shadowCasterTier']): number {
  switch (tier) {
    case 'primary':
      return 1;
    case 'secondary':
      return 0.85;
    case 'uncertain':
      return 0.65;
    case 'unknown':
    case undefined:
      return 0.7;
  }
}

function filterDecisionMultiplier(decision: Building['filterDecision']): number {
  switch (decision) {
    case 'include':
      return 1;
    case 'review':
      return 0.5;
    case 'exclude':
      return 0.3;
    case 'unknown':
    case undefined:
      return 0.75;
  }
}

function casterClassMultiplier(casterClass: Building['casterClass']): number {
  switch (casterClass) {
    case 'building':
      return 1;
    case 'manual_override':
      return 0.95;
    case 'structure':
      return 0.85;
    case 'vegetation':
      return 0.5;
    case 'unknown':
    case undefined:
      return 0.7;
  }
}

function clamp01(value: number): number {
  return Math.max(0.0, Math.min(1.0, value));
}

function extractPolygon(
  geometry: GeoJSON.Geometry
): GeoJSON.Polygon | null {
  if (geometry.type === 'Polygon') return geometry;
  if (geometry.type === 'MultiPolygon') {
    let largest: GeoJSON.Polygon | null = null;
    let maxArea = 0;
    for (const coords of geometry.coordinates) {
      const poly: GeoJSON.Polygon = { type: 'Polygon', coordinates: coords };
      const area = turf.area(turf.polygon(coords));
      if (area > maxArea) {
        maxArea = area;
        largest = poly;
      }
    }
    return largest;
  }
  return null;
}
